h2o-r/h2o-package/R/frame.R

```diff
@@ -3228,34 +3230,53 @@ as.h2o.Matrix <- function(x, destination_frame="", ...) {
   .key.validate(destination_frame)
   if ( destination_frame=="" ) # .h2o.readSVMLight wont handle ""
     destination_frame <- .key.make("Matrix") # only used if `x` variable name not valid key
-  
-  tmpf <- tempfile(fileext = ".svm")
-  .h2o.write.matrix.svmlight(x, file = tmpf)
-  h2f <- .h2o.readSVMLight(tmpf, destination_frame = destination_frame)
-  file.remove(tmpf)
-  h2f
+
+  drs = as.simple_triplet_matrix(x)# need to convert sparse matrix x to a simple triplet matrix format
+  thefile <- tempfile()
+  .h2o.write_stm_svm(drs, file = thefile)
+  h2f <- h2o.uploadFile(thefile, parse_type = "SVMLight")
+  unlink(thefile)
+  h2f[- c(1)]# remove the first column
+
 }
 
-.h2o.write.matrix.svmlight <- function(matrix, file) {
-  on.exit(sink())
-  sink(file)
-  sapply(1:nrow(matrix), function(i) {
-    r <- matrix[i, ]
-    val.indices <- which(r != 0)
-    val.indices <- val.indices[val.indices > 1]
-    target <- r[1]
-    features <- paste(sprintf("%d", val.indices - 1), r[val.indices], collapse = " ", sep = ":")
-    line <- sprintf("%s %s\n", target, features)
-    cat(line)
-  })
+#' Convert a simple triplet matrix to svm format
+#' @author Peter Ellis
+#' @return a character vector of length n = nrow(stm)
+.h2o.calc_stm_svm <- function(stm, y){
+  # returns a character vector of length y ready for writing in svm format
+  if(!"simple_triplet_matrix" %in% class(stm)){
+    stop("stm must be a simple triple matrix")
+  }
+  if(!is.vector(y) | nrow(stm) != length(y)){
+    stop("y should be a vector of length equal to number of rows of stm")
+  }
+  n <- length(y)
+
+
+  # data.table solution thanks to roland
+  rowLeft = setdiff(c(1:n), unique(stm$i))  # added two liner to return rows of zeros instead of repeating other rows
+  nrowLeft = length(rowLeft)
+
+  stm2 <- data.table(i = c(stm$i,rowLeft), j = c(stm$j,rep(1,nrowLeft)), v = c(stm$v,rep(0,nrowLeft)))
+  res <- stm2[, .(i, jv = paste(j, v, sep = ":"))][order(i), .(res = paste(jv, collapse = " ")), by = i][["res"]]
+
+  out <- paste(y, res)
+
+  return(out)
 }
 
-#'
-#' Converts parsed H2O data into an R data frame
-#'
-#' Downloads the H2O data and then scans it in to an R data frame.
-#'
-#' @param x An H2OFrame object.
+
+#' @param stm a simple triplet matrix (class exported slam) of features (ie explanatory variables)
```

Comment:
Hidden functions don't need roxygen headers. Let's remove those for consistency. If there are notes about the function and what it should produce, you can put that as a comment in the function itself.